# Assignment 3: Data

In [1]:
import requests

In [2]:
import json

In [3]:
sample_json = '''
{
  "station": "USC00305800",
  "name": "New York Central Park",
  "location": {
    "latitude": 40.7789,
    "longitude": -73.9692
  },
  "observations": [
    {"date": "2023-01-01", "temperature": 32, "precipitation": 0.0},
    {"date": "2023-01-02", "temperature": 28, "precipitation": 0.5},
    {"date": "2023-01-03", "temperature": 35, "precipitation": 0.0},
    {"date": "2023-01-04", "temperature": 38, "precipitation": 0.2},
    {"date": "2023-01-05", "temperature": 41, "precipitation": 0.0}
  ]
}
'''

data = json.loads(sample_json)

print("Station:", data['station'])
print("Location:", data['location'])
print("First observation:", data['observations'][0])


Station: USC00305800
Location: {'latitude': 40.7789, 'longitude': -73.9692}
First observation: {'date': '2023-01-01', 'temperature': 32, 'precipitation': 0.0}


In [31]:
# 1. Extract and print all dates and temperatures (8 points)
print("Date, Temperature")
for obs in data['observations']:
    print(f"{obs['date']}, {obs['temperature']}")

Date, Temperature
2023-01-01, 32
2023-01-02, 28
2023-01-03, 35
2023-01-04, 38
2023-01-05, 41


In [ ]:
# 2. Calculate average temperature (8 points)

In [54]:
total_temp = 0
count = 0
for obs in data['observations']:
    total_temp = total_temp + obs['temperature']
    count += 1
    
avg_temp =total_temp / count
print(f"Average temperature: {avg_temp}°F")

Average temperature: 34.8°F


In [ ]:
# 3. Find days with precipitation (9 points)

In [53]:
print("\nDays with precipitation:")
for obs in data['observations']:
    if obs['precipitation'] > 0:
        print (obs['date'])


Days with precipitation:
2023-01-02
2023-01-04


In [99]:
# Use a real weather API (you may need to sign up for a free API key)
# Example APIs: OpenWeatherMap, NOAA, Weather.gov
url= 'https://api.weather.gov/'
response = requests.get(url)

# Part 2: Downloading Files with Python

In [57]:
import pooch
import os

In [58]:
# Set up Pooch to download a file
# This example downloads a small air quality dataset
file_path = pooch.retrieve(
    url="https://github.com/pandas-dev/pandas/raw/main/doc/data/air_quality_no2.csv",
    known_hash=None
)

print("File downloaded to:", file_path)
print("File exists:", os.path.exists(file_path))

File downloaded to: /home/xm2357/.cache/pooch/458dad453f6a48e510cd544bef1854e3-air_quality_no2.csv
File exists: True


In [59]:
# 1. Verify the file was downloaded (5 points)
# Check the file size
file_size = os.path.getsize(air_quality_no2.csv)
print(f"File size: {file_size} bytes")

File size: 31984 bytes


In [61]:
import pandas as pd

In [80]:
# YOUR CODE HERE: open the file and count how many lines it has
df = pd.read_csv(file_path)
df

datetime  station_antwerp  station_paris  station_london
0     2019-05-07 02:00:00              NaN            NaN            23.0
1     2019-05-07 03:00:00             50.5           25.0            19.0
2     2019-05-07 04:00:00             45.0           27.7            19.0
3     2019-05-07 05:00:00              NaN           50.4            16.0
4     2019-05-07 06:00:00              NaN           61.9             NaN
...                   ...              ...            ...             ...
1030  2019-06-20 22:00:00              NaN           21.4             NaN
1031  2019-06-20 23:00:00              NaN           24.9             NaN
1032  2019-06-21 00:00:00              NaN           26.5             NaN
1033  2019-06-21 01:00:00              NaN           21.8             NaN
1034  2019-06-21 02:00:00              NaN           20.0             NaN

[1035 rows x 4 columns]

In [82]:
line_count = len(df) 
print(f"Number of lines: {line_count}")

Number of lines: 1035


In [83]:
# 2. Download another file (10 points)
# Find a climate dataset online using the sources we talked about in lecture
# Download it using Pooch
# YOUR CODE HERE:
my_url = "https://www.cpc.ncep.noaa.gov/data/indices/ersst5.nino.mth.81-10.ascii"
my_file = pooch.retrieve(url=my_url, known_hash=None)

SHA256 hash of downloaded file: 4c9bdf8ffe8e9501f0688cb96f6d5bed8ab91ed3d095801d0be61e5bd2286ec3
Use this value as the 'known_hash' argument of 'pooch.retrieve' to ensure that the file hasn't changed if it is downloaded again in the future.


In [84]:
print("File downloaded to:", my_file)
print("File exists:", os.path.exists(my_file))

File downloaded to: /home/xm2357/.cache/pooch/64081e4fd24be107c751f569733fc165-ersst5.nino.mth.81-10.ascii
File exists: True


In [89]:
df2=pd.read_csv(my_file,delim_whitespace=True)

/tmp/ipykernel_360236/3783962422.py:1: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  df2=pd.read_csv(my_file,delim_whitespace=True)


In [88]:
df2

YR  MON  NINO1+2  ANOM  NINO3  ANOM.1  NINO4  ANOM.2  NINO3.4  ANOM.3
0    1950    1    23.01 -1.53  23.56   -2.11  26.94   -1.30    24.55   -2.00
1    1950    2    24.32 -1.70  24.89   -1.52  26.67   -1.43    25.06   -1.68
2    1950    3    25.11 -1.31  26.36   -0.80  26.52   -1.68    25.87   -1.37
3    1950    4    23.63 -1.83  26.44   -1.04  26.90   -1.55    26.28   -1.43
4    1950    5    22.68 -1.59  25.69   -1.47  27.73   -1.01    26.18   -1.63
..    ...  ...      ...   ...    ...     ...    ...     ...      ...     ...
847  2020    8    19.96 -1.05  24.50   -0.60  28.48   -0.17    26.27   -0.57
848  2020    9    19.50 -1.23  23.91   -1.02  28.21   -0.45    25.89   -0.84
849  2020   10    20.25 -0.80  23.85   -1.10  27.96   -0.68    25.46   -1.21
850  2020   11    20.95 -0.75  23.89   -1.15  27.81   -0.75    25.29   -1.33
851  2020   12    22.03 -0.86  24.33   -0.87  27.55   -0.89    25.42   -1.14

[852 rows x 10 columns]

In [90]:
# 3. Create a data inventory (5 points)
# List all the files you've downloaded in this assignment
print("\nData Inventory:")
print("1. meteorites.csv - NASA meteorite landings")
print("2. air_quality_no2.csv - Air quality NO2 measurements")
print("3. ersst5.nino.mth.81-10.ascii - El Niño data")


Data Inventory:
1. meteorites.csv - NASA meteorite landings
2. air_quality_no2.csv - Air quality NO2 measurements
3. ersst5.nino.mth.81-10.ascii - El Niño data


# Part3: Understanding NetCDF Metadata

In [91]:
import requests

In [93]:
# OPeNDAP provides metadata in different formats
# We'll get basic info about a climate dataset

base_url = "http://iridl.ldeo.columbia.edu/expert/SOURCES/.NOAA/.NCEP/.CPC/.UNIFIED_PRCP/.GAUGE_BASED/.GLOBAL/.v1p0/.Monthly/.RETRO/.rain/dods"

# Get DDS (Dataset Descriptor Structure) - describes the structure
dds_url = base_url + ".dds"
response = requests.get(dds_url)

print("Dataset Structure:")
print(response.text[:500])  # Print first 500 characters

Dataset Structure:
Dataset {
    Float32 Y[Y = 360];
    Float32 X[X = 720];
    Float32 T[T = 324];
    Grid {
     ARRAY:
        Float32 rain[T = 324][Y = 360][X = 720];
     MAPS:
        Float32 T[T = 324];
        Float32 Y[Y = 360];
        Float32 X[X = 720];
    } rain;
} rain;



1. Identify dimensions and variables (5 points)
Look at the DDS output above and answer:
- What are the dimension names?
Y, X, T
- What is the main variable name?
rain
- Write your answers in a markdown cell

In [97]:
# 2. Get data attributes (5 points)
# DAS (Dataset Attribute Structure) contains metadata
das_url = base_url + ".das"
# YOUR CODE HERE: make a request to das_url and print first 1000 characters
response = requests.get(das_url)
print("Dataset Structure:")
print(response.text[:1000])

Dataset Structure:
Attributes {
    X {
        String standard_name "longitude";
        Float32 pointwidth 0.5;
        Int32 gridtype 1;
        String units "degree_east";
    }
    T {
        Float32 pointwidth 1.0;
        String calendar "360";
        Int32 gridtype 0;
        String units "months since 1960-01-01";
    }
    Y {
        String standard_name "latitude";
        Float32 pointwidth 0.5;
        Int32 gridtype 0;
        String units "degree_north";
    }
    rain {
        Int32 pointwidth 0;
        String standard_name "lwe_precipitation_rate";
        Float32 file_missing_value -999.0;
        String history "Boxes with less than 0.0% dropped";
        Float32 missing_value NaN;
        String units "mm/day";
        String long_name "Monthly Precipitation";
    }
NC_GLOBAL {
    String Conventions "IRIDL";
}
}



3. Document what you learned (5 points)
In a markdown cell, write:
- What does this dataset contain?
latitude, longitude, time(month/year), monthly precipitation

- What time period does it cover?
1960-01-01 + 324months: so it covers from 1960-01-01 to 1986-12-31.

- What geographic region does it cover?
x longitude, degree_east, pointwidth 0.5, X = 720
y latitude, degree_north, pointwidth 0.5, Y = 360
it covers longitude 360 degree and latitude 180 degree.It is the globe. 

- What are the units of the main variable?
mm/day
Find this info in the DAS output
